sdk/opendp/whitenoise/synthesizers/dpgan/dpgan.py

```diff
@@ -0,0 +1,184 @@
+import torch.nn.functional as F
+import torch.optim as optim
+from torch.utils.data import DataLoader, TensorDataset
+
+from torchdp import PrivacyEngine, utils, autograd_grad_sample
+from .utils import GeneralTransformer
+
+from opendp.whitenoise.synthesizers.base import SDGYMBaseSynthesizer
+
+class Generator(nn.Module):
+    def __init__(self, latent_dim, hidden_dims, output_dim):
+        super(Generator, self).__init__()
+        
+        hidden_activation = nn.ReLU()
+        previous_layer_dim = latent_dim
+        
+        model = []
+        for i, layer_dim in enumerate(list(hidden_dims)):
+            model.append(nn.Linear(previous_layer_dim, layer_dim))
+            model.append(nn.BatchNorm1d(layer_dim))
+            model.append(hidden_activation)
+            previous_layer_dim = layer_dim
+        model.append(nn.Linear(previous_layer_dim, output_dim))
+        
+        self.model = nn.Sequential(*model)
+    
+    def forward(self, noise):
+        data = self.model(noise)
+        return data
+
+class Discriminator(nn.Module):
+    def __init__(self, input_dim, hidden_dims, pac=4):
+        super(Discriminator, self).__init__()
+        dim = input_dim * pac
+        self.pac = pac
+        self.pac_dim = dim
+        
+        model = []
+        for i, layer_dim in enumerate(list(hidden_dims)):
+            model.append(nn.Linear(dim, layer_dim))
+            model.append(nn.LeakyReLU(0.2))
+            model.append(nn.Dropout(0.5))
+            dim = layer_dim
+        model.append(nn.Linear(dim, 1))
+        self.model = nn.Sequential(*model)
+    
+    def forward(self, input):
+        return self.model(input.view(-1, self.pac_dim))
+
+def apply_activate(data, output_info):
+    data_t = []
+    st = 0
+    for item in output_info:
+        if item[1] == 'tanh':
+            ed = st + item[0]
+            data_t.append(torch.tanh(data[:, st:ed]))
+            st = ed
+        elif item[1] == 'softmax':
+            ed = st + item[0]
+            data_t.append(F.gumbel_softmax(data[:, st:ed], tau=0.2))
+            st = ed
+        else:
+            continue
+    return torch.cat(data_t, dim=1)
+
+class DPGANSynthesizer(SDGYMBaseSynthesizer):
+    def __init__(self,
+                 latent_dim=128,
+                 gen_dim=(256, 256),
+                 dis_dim=(256, 256),
+                 batch_size=256,
+                 epochs=1000,
+                 n_critics=5, 
+                 delta=1e-5,
+                 budget=3.0):
+        self.latent_dim = latent_dim
+        self.gen_dim = gen_dim
+        self.dis_dim = dis_dim
+        self.batch_size = batch_size
+        self.epochs = epochs
+        self.n_critics = n_critics
+        self.delta = delta
+        self.budget = budget
+        
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    
+    def fit(self, data, categorical_columns=tuple(), ordinal_columns=tuple()):
+        self.transformer = GeneralTransformer()
```

Comment:
@eedeleon this line here is what I'm referring to. Here we instantiate the object that creates a vector embedding from the original pandas dataframe which then gets passed to the discriminator network. The vector embedding might change though if someone uses an autoencoder or some other preprocessing step for example